signer/xpi/jar.go

```diff
@@ -7,20 +7,58 @@ import (
 	"crypto/sha256"
 	"encoding/base64"
 	"fmt"
+	"github.com/pkg/errors"
 	"io"
 	"io/ioutil"
 	"strings"
-	"github.com/pkg/errors"
 )
 
-func makeJARManifests(input []byte) (manifest, sigfile []byte, err error) {
+func makePKCS7Manifest(input []byte, metafiles []Metafile) (manifest []byte, err error) {
+	manifest, err = makeJARManifest(input)
+	if err != nil {
+		return
+	}
+
+	mw := bytes.NewBuffer(manifest)
+	for _, f := range metafiles {
+		fmt.Fprintf(mw, "Name: %s\nDigest-Algorithms: SHA1 SHA256\n", f.Name)
+		h1 := sha1.New()
+		h1.Write(f.Body)
+		fmt.Fprintf(mw, "SHA1-Digest: %s\n", base64.StdEncoding.EncodeToString(h1.Sum(nil)))
+		h2 := sha256.New()
+		h2.Write(f.Body)
+		fmt.Fprintf(mw, "SHA256-Digest: %s\n\n", base64.StdEncoding.EncodeToString(h2.Sum(nil)))
+	}
+
+	return mw.Bytes(), err
+}
+
+// makeJARManifestAndSignature writes hashes for all entries in a zip to a
+// manifest file then hashes the manifest file to write a signature
+// file and returns both
+func makeJARManifestAndSignature(input []byte) (manifest, sigfile []byte, err error) {
+	manifest, err = makeJARManifest(input)
+	if err != nil {
+		return
+	}
+
+	sigfile, err = makeJARSignature(manifest)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// makeJARManifest calculates a sha1 and sha256 hash for each zip entry and writes them to a manifest file
+func makeJARManifest(input []byte) (manifest []byte, err error) {
 	inputReader := bytes.NewReader(input)
 	r, err := zip.NewReader(inputReader, int64(len(input)))
 	if err != nil {
 		return
 	}
 
-	// first generate the manifest file by calculating a sha1 and sha256 hash for each zip entry
+	// generate the manifest file by calculating a sha1 and sha256 hash for each zip entry
```

Comment:
nit: `isSignatureFile` further down should probably be renamed to `isJARSignatureFile` since it doesn't check for COSE signature files.